### PR TITLE
Feat/validate update device completed

### PR DESCRIPTION
**Pull request by maximsk**
Add new parser: `UpdateDeviceCompletedSystemLogMessage.java`
Documentation: https://www.bsi.bund.de/SharedDocs/Downloads/DE/BSI/Publikationen/TechnischeRichtlinien/TR03153/TR-03153_Anwendungshinweise.pdf?__blob=publicationFile&v=1

### CODE REVIEW

**Review by williambaer — APPROVED** [review]

File: src/main/resources/ttc.properties [hunk 1]
```diff
@@ -94,6 +94,16 @@ de.konfidas.ttc.messages.systemlogs.logoutCauseNotFound = Fehler beim Parsen des
...
+de.konfidas.ttc.messages.systemlogs.errorParsingSystemOperationDataNewVersionNotFound = Fehler beim Parsen des systemOperationDataContent. Das Pflichtfeld newVersion wurde nicht gefunden
+de.konfidas.ttc.messages.systemlogs.errorParsingSystemOperationDataComponentNameNotFound = Fehler beim Parsen des systemOperationDataContent. Das Pflichtfeld componentName wurde nicht gefunden
+de.konfidas.ttc.messages.systemlogs.errorParsingSystemOperationDataManufacturerNotFound = Fehler beim Parsen des systemOperationDataContent. Das Pflichtfeld manufacturer wurde nicht gefunden
+de.konfidas.ttc.messages.systemlogs.errorParsingSystemOperationDataModelNotFound = Fehler beim Parsen des systemOperationDataContent. Das Pflichtfeld model wurde nicht gefunden
+de.konfidas.ttc.messages.systemlogs.errorParsingSystemOperationDataVersionNotFound = Fehler beim Parsen des systemOperationDataContent. Das Pflichtfeld version wurde nicht gefunden
+de.konfidas.ttc.messages.systemlogs.errorParsingSystemOperationDataCertificationIdNotFound = Fehler beim Parsen des systemOperationDataContent. Das Pflichtfeld certificationId wurde nicht gefunden
```
Comment (kplakidas) [inline src/main/resources/ttc.properties]: certificationId is optional, so why would there be such an error?

File: src/main/java/de/konfidas/ttc/messages/systemlogs/UpdateDeviceCompletedSystemLogMessage.java [hunk 2]
```diff
@@ -13,121 +13,200 @@
...
+            this.allErrors.add(new SystemLogParsingError(properties.getString("de.konfidas.ttc.messages.systemlogs.errorParsingSystemOperationDataDeviceInformationSetTooShort")));
+        }
 
+        this.deviceInformationSet.forEach(component -> {
+            if (!this.requiredComponentNames.contains(component.componentName)) {
+                this.allErrors.add(new SystemLogParsingError(properties.getString("de.konfidas.ttc.messages.systemlogs.errorParsingSystemOperationDataComponentMissing")));
+            }
+        });
+    }
 }
```
Comment (kplakidas) [inline src/main/java/de/konfidas/ttc/messages/systemlogs/UpdateDeviceCompletedSystemLogMessage.java]: newline missing

File: src/main/resources/ttc.properties (hunk 1 above)
Comment (maximsk) [reply src/main/resources/ttc.properties]: thanks, error message changed.

File: src/main/java/de/konfidas/ttc/messages/systemlogs/UpdateDeviceCompletedSystemLogMessage.java [hunk 3]
```diff
@@ -13,121 +13,205 @@
...
 
-            this.oldVersion = (DLTaggedObject) systemOperationDataIterator.next();
+            //read reasonForFailure
+            if (nextElement.getTagNo() == 2) {
+                this.reasonForFailure = (DLTaggedObject) systemOperationDataIterator.next();
+                this.reasonForFailureAsString = DLTaggedObjectConverter.dLTaggedObjectToString(this.reasonForFailure);
+            }
 
-            //oldVersion is a Sequence in itself. Parsing follows
+            nextElement = (DLTaggedObject) systemOperationDataAsAsn1List.get(systemOperationDataIterator.nextIndex());
```
Comment (kplakidas) [inline src/main/java/de/konfidas/ttc/messages/systemlogs/UpdateDeviceCompletedSystemLogMessage.java:111]: What happens here if the previous `nextElement` was already the `newVersion` field (tag 3)? Won't the `nextElement` assignment or the following if clause result in error?

Comment (maximsk) [reply src/main/java/de/konfidas/ttc/messages/systemlogs/UpdateDeviceCompletedSystemLogMessage.java:111]: Nothing. As far as iterator doesn't change the position in this case, there is no problem.

**Pull request merged**